### PR TITLE
arch/arm/soc/nxp_imx: added support to imx7d m4 core

### PR DESCRIPTION
**Pull request by uLipe**
board/colibri_imx7d_m4: added board support to toradex colibri imx7d single board computer

This change basically adds support of a popular IMX7D processors from NXP, which has a ARM Cortex A7 Core and a Cortex M4 Core, this change is targeted to the second one and now is possible to select it as a supported SoC, Toradex has some attractive single board computers with industrial grade using this processor and this change also the basic support of the board and the NXP supplied HAL.

This support is intended to increase the options of heterogeneous processors in Zephyr SoC portfolio, adds the SoC and Board template is the first step only, so I divided the change in two steps, first provide a functional SoC support to perform build and image generation to be flashed to TCMRAM (imx7d m4 core does not has built in flash), the next step is to provide the serial drivers and the IPM driver to communicate m4 core with a7 core, this approach was chosen to keep the changes not too large and hard to verify.

This changed was validated as follows:
 - built the hello world sample under build folder with: $cmake -DBOARD=colibri_imx7d_m4 .. and make.
 - the generated image can be loaded through a Jlink probe or with A7 running Linux + uBoot using the fatload command

Signed-off-by: Felipe Neves <ryukokki.felipe@gmail.com>

### CODE REVIEW

**Review by galak — CHANGES_REQUESTED** [review]
Can you split this into at least 2 patches, one for the soc code and one for the board.

**Comment by uLipe** [comment]
Hi, yes I can. 


In this case I will modify this patch and the next one include board support.

Thanks

**Comment by pfalcon** [comment]
@uLipe : Note that you can just make this pull request contain 2 commits (no need to create 2 pull requests for that).

**Comment by uLipe** [comment]
@pfalcon, thanks for the tip, I will submit the just one PR containing the soc and board commits.

Best :)

File: boards/arm/colibri_imx7d_m4/colibri_imx7d_m4_defconfig [hunk 1]
```diff
@@ -0,0 +1,15 @@
+CONFIG_ARM=y
+CONFIG_SOC_FAMILY_IMX=y
+CONFIG_SOC_SERIES_IMX7_M4=y
+CONFIG_SOC_MCIMX7_M4=y
+CONFIG_BOARD_COLIBRI_IMX7D_M4=y
+CONFIG_CORTEX_M_SYSTICK=y
+CONFIG_UART_NXP_IMX=y
```
Comment (MaureenHelm) [inline boards/arm/colibri_imx7d_m4/colibri_imx7d_m4_defconfig]: Remove this line, we can select the appropriate uart driver in the SoC defconfig. See arch/arm/soc/nxp_imx/rt/Kconfig.defconfig.mimxrt1052 for an example.

File: dts/arm/nxp/nxp_imx.dtsi [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+		cpu@0 {
+			device_type = "cpu";
+			compatible = "arm,cortex-m4";
+			reg = <0>;
+		};
+	};
+
+	soc {
+
+		tcram: tcram@1fff8000 {
```
Comment (MaureenHelm) [inline dts/arm/nxp/nxp_imx.dtsi]: Let's call this tcml to match the hardware reference manual, and add tcmu.

File: dts/arm/nxp/nxp_imx.dtsi [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+			clock-controller;
+			#clocks-cells = <3>;
+		};
+
+
+		ddram: ddram@10000000 {
+			compatible = "nxp,imx-ddram";
+			reg = <0x10000000 0x10000000>;
+			label = "CCM";
```
Comment (MaureenHelm) [inline dts/arm/nxp/nxp_imx.dtsi]: s/CCM/DDR/

File: dts/arm/nxp/nxp_imx.dtsi [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+		iomuxc: iomuxc@30330000 {
+			reg = <0x30330000 0x4000>;
+			label = "PINMUX_0";
+		};
+
+
+		uart2: uart@30890000 {
+			compatible = "nxp,imx-uart";
+			reg = <0x30890000 0x4000>;
```
Comment (MaureenHelm) [inline dts/arm/nxp/nxp_imx.dtsi]: Base address should be 0x30870000

File: ext/hal/Kconfig [hunk 5]
```diff
@@ -20,6 +20,8 @@ source "ext/hal/nordic/Kconfig"
 
 source "ext/hal/nxp/mcux/Kconfig"
 
+source "ext/hal/nxp/imx/Kconfig"
```
Comment (MaureenHelm) [inline ext/hal/Kconfig]: Merge contents into existing ext/hal/nxp/mcux

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Hi @uLipe , thanks for the PR! This is something we have wanted to do for some time.

As others have already said, please split this up into multiple commits as it will help with the review process and keep the git history clean. It's ok to force push your branch to update the PR. I think you should have three commits, rather than two:
1. Import mcux hal for imx7d. Please reuse the existing ext/hal/nxp/mcux folder and configs rather than creating a new folder and configs.
2. Add imx7d SoC support
3. Add colibri board support. Please also add a board doc to this commit, refer to boards/arm/mimxrt1050_evk/doc/mimxrt1050_evk.rst as an example.

cc: @MarekNovakNXP

**Comment by uLipe** [comment]
@MaureenHelm , thanks for the tips I started to work in these changes, let me ask a question about the mcux hal and imx hal, I noted the nxp sdk for the IMX7 microcontroller side seems to not compatible with the new mcuxpresso drivers (the files are quite different and the IPs too), more than that currently the MCUX dashboard does not support the imx7d processors, the duplication of the hal was based on this incompatibility, but I agree with you to keep only the mcux hal.

Given this scenario the question is: Can I move the drivers source files of IMX7 directly to the mcuexpresso drivers/ directory? or should I create a second directory called for example: drivers_imx7d ?

**Comment by MaureenHelm** [comment]
> I noted the nxp sdk for the IMX7 microcontroller side seems to not compatible with the new mcuxpresso drivers (the files are quite different and the IPs too), more than that currently the MCUX dashboard does not support the imx7d processors, the duplication of the hal was based on this incompatibility, but I agree with you to keep only the mcux hal.

My mistake, I thought we had imx7d m4 support in mcux 2.x but it turns out the drivers and headers are based on ksdk 1.x. Let me check with the internal development team to see if there are any plans to update it, and then we can decide how to proceed.

**Comment by galak** [comment]
@MaureenHelm curious if the imx7d and imx6 code would be similar (for those imx6 devices with cortex-m as well).

**Comment by MaureenHelm** [comment]
> @MaureenHelm curious if the imx7d and imx6 code would be similar (for those imx6 devices with cortex-m as well).

For now, yes. Existing imx6 and imx7 devices have similar ksdk 1.x based m4 drivers, but future devices in these families may have mcux 2.x. 


> My mistake, I thought we had imx7d m4 support in mcux 2.x but it turns out the drivers and headers are based on ksdk 1.x. Let me check with the internal development team to see if there are any plans to update it, and then we can decide how to proceed.

There are no plans to update this device to mcux 2.x, so I now agree with your original decision to create a new hal folder. I'm not sure I like "ext/hal/nxp/imx" and CONFIG_HAS_IMX_HAL though, as I think that will create confusion for imx devices that do have mcux. NXP releases the code in a "FreeRTOS BSP", which is based on ksdk 1.x, so some possible alternatives are:
ext/hal/nxp/freertos_bsp
ext/hal/nxp/ksdk

Thoughts?

File: .gitignore [hunk 6]
```diff
@@ -7,6 +7,7 @@
 *.swp
 *.swo
 *~
+*.DS_Store
```
Comment (galak) [inline .gitignore]: I'm not really sure we should be adding this, and if it should be in a separate PR/commit.

File: samples/net/sockets/dumb_http_server/src/response_big.html.bin.inc [hunk 7]
```diff
@@ -0,0 +1,273 @@
+0x48, 0x54, 0x54, 0x50, 0x2f, 0x31, 0x2e, 0x30,
```
Comment (galak) [inline samples/net/sockets/dumb_http_server/src/response_big.html.bin.inc]: This file should not be check in.

**Review by galak — CHANGES_REQUESTED** [review]
Remove the .bin.inc files that are checked in.

**Comment by uLipe** [comment]
Hi,

@MaureenHelm, thanks for the support, I agree the imx named folder in hal could cause confusion, ksdk seems confusing too since it refers to the old kinetis sdk.

freertos_bsp is also too generic but since it will contained in nxp folder, this can be a solution.

Should i proceed with hal/nxp/freertos_sdk(or bsp)?

Thanks again :)

**Comment by galak** [comment]
> There are no plans to update this device to mcux 2.x, so I now agree with your original decision to create a new hal folder. I'm not sure I like "ext/hal/nxp/imx" and CONFIG_HAS_IMX_HAL though, as I think that will create confusion for imx devices that do have mcux. NXP releases the code in a "FreeRTOS BSP", which is based on ksdk 1.x, so some possible alternatives are:
> ext/hal/nxp/freertos_bsp
> ext/hal/nxp/ksdk
> 
> Thoughts?

Do you expect that something like i.mx8 would be supported by mcux?

I wondering if we should just have native drivers.  Wondering how many peripheral drivers would actually be used/supported?  @uLipe which devices do you see using from the Cortex-M4 on i.mx7d?

**Comment by MaureenHelm** [comment]
> Do you expect that something like i.mx8 would be supported by mcux?

Yes

> I wondering if we should just have native drivers.

Why? We still need the device headers, so doing native drivers doesn't eliminate the problem of how to import the nxp freertos bsp.

> Should i proceed with hal/nxp/freertos_sdk(or bsp)

I think we should go with freertos_bsp. It's a little weird, but it's closest to the package name from nxp. @galak, do you agree?

**Comment by galak** [comment]
> I think we should go with freertos_bsp. It's a little weird, but it's closest to the package name from nxp. @galak, do you agree?

Uugh, why'd you guys have to use freertos in the bsp name ;).  Could we do something like ext/hal/nxp/imx_legacy/ . or imx_6_7 or something like that to convey the SoCs we are supporting?

**Comment by MaureenHelm** [comment]
> Uugh, why'd you guys have to use freertos in the bsp name ;). Could we do something like ext/hal/nxp/imx_legacy/ or 'imx_6_7' or something like that to convey the SoCs we are supporting?

No, there may be new SoCs in these families that have mcux. I don't like having freertos in the name either, but I still think keeping the folder name close to the nxp package name makes the origin clear.

**Comment by galak** [comment]
@uLipe - any feedback to my question?

**Comment by uLipe** [comment]
Hello all, sorry for long delayed response (I was travelling with very limited internet access).

Regarding the question, @galak , I don't think the mention of freertos is the more correct way, but based on sdk delivered by NXP and if the goal is to make the external resources on hal folder compliant with the chip manufacturer, the freertos_bsp is the better way to naming,using imx_legacy is not good at all too because IMX6 and 7 families as well their sdk are not obsolete, discontinued or even legacy, is just not compatible with the current mcux sdk. 

I used the name as just "imx" because it sounds a mid-term but at cost of some confusion risk by the user, @MaureenHelm can I keep the naming of the original PR? I think, because mcux incompatibility by the IMX7 SoCs, the naming solution chosen will aways have a drawback.

I will resume the PR task at this weekend. 

Thanks again by the support.

Felipe

**Comment by galak** [comment]
@uLipe any chance you'll get back to this?

**Comment by uLipe** [comment]
Hello @galak , I will suffering some busy times here and due this I will not be able to continue it, my colleague @diegosueiro will take this task from ahead.

Thanks for the support

**Comment by MaureenHelm** [comment]
> I used the name as just "imx" because it sounds a mid-term but at cost of some confusion risk by the user, @MaureenHelm can I keep the naming of the original PR? I think, because mcux incompatibility by the IMX7 SoCs, the naming solution chosen will aways have a drawback.

Yes

> I will suffering some busy times here and due this I will not be able to continue it, my colleague @diegosueiro will take this task from ahead.

Understand. Thanks @diegosueiro for picking this up.

**Comment by diegosueiro** [comment]
@MaureenHelm, 

It will be a pleasure to continue this PR.
Actually, I'm the original author of this porting and @uLipe helped me to adapt it to use cmake (when I started Zephyr was based on KBuild), among other things and created the PR.

I'm just wondering if it is better to create a new PR applying the changes suggested by you and @galak.

I'm just confused regarding the naming the hal directory. What should I use? "ext/hal/nxp/imx"?

**Comment by MaureenHelm** [comment]
> I'm just wondering if it is better to create a new PR applying the changes suggested by you and @galak.

Either way is fine. We usually do force pushes to update a PR, but in this case you would have to force push to @uLipe 's repo. If you do create a new PR, please add a reference to this one in the description.

> I'm just confused regarding the naming the hal directory. What should I use? "ext/hal/nxp/imx"?

Yes, sorry for the confusion. I don't really like this name or any of the others we kicked around, but I don't have any better suggestions. Please also have a look at http://docs.zephyrproject.org/contribute/contribute_non-apache.html

**Comment by diegosueiro** [comment]
@MaureenHelm,

> Either way is fine. We usually do force pushes to update a PR, but in this case you would have to force push to @uLipe 's repo. If you do create a new PR, please add a reference to this one in the description.

Got it. I'll create a new PR.

> Please also have a look at http://docs.zephyrproject.org/contribute/contribute_non-apache.html

The iMX7D drivers and utilities were taken from NXP FreeRTOS BSP and according to the file "SW-Content-Register-FreeRTOS-BSP-1.0.1-i.MX7D.txt" they are released under BSD-3-Clause.

In this case, should I need to make the submission for TSC review?

**Comment by MaureenHelm** [comment]
> In this case, should I need to make the submission for TSC review?

Please add a README to your PR, and I'll submit it to the TSC/gov board for review.

**Comment by diegosueiro** [comment]
@MaureenHelm,

Just to let you know that I just submitted the PR #6367

**Comment by MaureenHelm** [comment]
> Just to let you know that I just submitted the PR #6367

Thanks!

**Pull request closed without merging**